drgn_tools/lock.py

```diff
@@ -128,6 +216,26 @@ def scan_sem_lock(prog: Program, stack: bool) -> None:
         show_sem_lock(prog, frame_list, seen_sems, stack)
 
 
+def scan_rwsem_lock(prog: Program, stack: bool) -> None:
+    """Scan for read-write(rw) semphores"""
+    seen_rwsems: Set[int] = set()
+    frame_list = bt_has(prog, "__rwsem_down_write_failed_common")
+    if frame_list:
+        show_rwsem_lock(prog, frame_list, seen_rwsems, stack)
+
+    frame_list = bt_has(prog, "__rwsem_down_read_failed_common")
+    if frame_list:
+        show_rwsem_lock(prog, frame_list, seen_rwsems, stack)
+
+    frame_list = bt_has(prog, "rwsem_down_write_slowpath")
+    if frame_list:
+        show_rwsem_lock(prog, frame_list, seen_rwsems, stack)
+
+    frame_list = bt_has(prog, "rwsem_down_read_slowpath")
+    if frame_list:
+        show_rwsem_lock(prog, frame_list, seen_rwsems, stack)
```

Comment:
True. I also found it unoptimal while testing vmcores locally. bt_has was taking few secs but I went ahead with what was available at that point of time. I had some optimisations in mind, mentioning them here as you are already working on it:
1. For scanning waiters we just need to traverse tasks that are in blocked state, may be possible once my other PR for traversing task in different states is available
2. have a version of bt_has that can be used with a given task_struct. that way we could cache tasks that are waiting on X and check from remaining the ones waiting on Y

Reply:
The biggest issue is that multiple calls to `bt_has()` will re-run the search. So the biggest performance win would be a `bt_has_any(prog, ["function1", "function2"])` rather than calling `bt_has()` several times. In fact, my current optimization simply caches a mapping of each function name to the tasks which contain it, for all functions that are on any stack. This isn't much more expensive than searching for a single function, and it makes repeated calls cheaper.

From there, your optimizations could be helpful as well. I pushed my suggestions in #72, let me know what you think and I'll continue looking at this.

Reply:
Sounds good. I will check and update you. thanks for starting on it quickly , will be a big help for sure.